### PR TITLE
xtensa: add support for platforms with no threadptr register.

### PR DESCRIPTION
**Pull request by lgirdwood**
Provide a virtual threadptr register in the memory map using a linker
area below the stack. This can then be referenced instead of the register
for ISAs that dont support threadptr.

This patch also includes initial support for HSW/BDW platforms.

TODO: This patch now allow BDW to boot on real HW and updated qemu, but it fails on any context switch to IRQ level1 ? i.e. scheduling IRQs seem to work but any IPC (level1) will process and then break next scheduler IRQ invocation (causing a kernel exception). @tlauda any ideas ? 

Signed-off-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>

### CODE REVIEW

File: src/platform/haswell/include/arch/xtensa/config/core-isa-bdw.h [hunk 1]
```diff
@@ -89,7 +89,7 @@
 #define XCHAL_HAVE_PSO			0	/* Power Shut-Off */
 #define XCHAL_HAVE_PSO_CDM		0	/* core/debug/mem pwr domains */
 #define XCHAL_HAVE_PSO_FULL_RETENTION	0	/* all regs preserved on PSO */
-#define XCHAL_HAVE_THREADPTR		1	/* THREADPTR register */
+#define XCHAL_HAVE_THREADPTR		0	/* THREADPTR register */
```
Comment (tlauda) [inline src/platform/haswell/include/arch/xtensa/config/core-isa-bdw.h]: I think we should update the whole header.
There is a high probability, that the rest of the values is also wrong.

File: src/platform/haswell/include/platform/lib/memory.h [hunk 2]
```diff
@@ -133,7 +135,11 @@ void platform_init_memmap(void);
 #define SOF_STACK_BASE		(DRAM0_BASE + DRAM0_SIZE)
 #define SOF_STACK_END		(SOF_STACK_BASE - SOF_STACK_TOTAL_SIZE)
 
-#define MAILBOX_BASE			(SOF_STACK_END - MAILBOX_SIZE)
+/* Virtual threadptr */
+#define SOF_VIRTUAL_THREAD_SIZE	0x20
```
Comment (tlauda) [inline src/platform/haswell/include/platform/lib/memory.h]: 4 bytes would be enough, but I assume it's some kind of precaution?

**Comment by tlauda** [comment]
@lgirdwood Hard to tell from the code what can be wrong. If your macro loads the right address for the virtual threadptr, then it should work correctly.

File: src/platform/haswell/include/arch/xtensa/config/core-isa-bdw.h (hunk 1 above)
Comment (lgirdwood) [reply src/platform/haswell/include/arch/xtensa/config/core-isa-bdw.h]: I checked the others and they looked correct.

File: src/platform/haswell/include/platform/lib/memory.h (hunk 2 above)
Comment (lgirdwood) [reply src/platform/haswell/include/platform/lib/memory.h]: Alignment and support for more cores.

**Comment by lgirdwood** [comment]
> @lgirdwood Hard to tell from the code what can be wrong. If your macro loads the right address for the virtual threadptr, then it should work correctly.

Yeah, validated on qemu. My suspicion is that level 1 IRQs are not used anywhere except on HSW/BDW since they have a different xtos entry compared to mid-high level IRQs (2 - 7).

**Comment by tlauda** [comment]
> > @lgirdwood Hard to tell from the code what can be wrong. If your macro loads the right address for the virtual threadptr, then it should work correctly.
> 
> Yeah, validated on qemu. My suspicion is that level 1 IRQs are not used anywhere except on HSW/BDW since they have a different xtos entry compared to mid-high level IRQs (2 - 7).

BYT/CHT uses level 1 as `PLATFORM_SCHEDULE_IRQ`. I see in CI that HSW fails and BDW boots OK. Is this expected, or maybe the exception you're seeing is not on BDW?

**Comment by lgirdwood** [comment]
> > > @lgirdwood Hard to tell from the code what can be wrong. If your macro loads the right address for the virtual threadptr, then it should work correctly.
> > 
> > 
> > Yeah, validated on qemu. My suspicion is that level 1 IRQs are not used anywhere except on HSW/BDW since they have a different xtos entry compared to mid-high level IRQs (2 - 7).
> 
> BYT/CHT uses level 1 as `PLATFORM_SCHEDULE_IRQ`. 
Ok, that's good to know since the entry is different.

I see in CI that HSW fails and BDW boots OK. Is this expected, or maybe the exception you're seeing is not on BDW?

Expected. Qemu has the wrong ISA config.

**Comment by lgirdwood** [comment]
@tlauda updated core-isa.h fixes the issue. FW can now boot again on BDW and perform IPC. Now need to update kernel and topology for XPS13. Good to merge.

**Comment by tlauda** [comment]
@lgirdwood What with the HSW boot fail? I think we should first update QEMU in CI to avoid constant failures.

**Comment by lgirdwood** [comment]
@tlauda yep, I will be asking @xiulipan to update qemu shortly. once this is done we can merge this PR.

**Comment by lgirdwood** [comment]
@xiulipan Travis will need an Qemu update now that https://github.com/thesofproject/qemu/pull/19 is merged. This fixes all HSW/BDW issues.

**Comment by xiulipan** [comment]
@lgirdwood Let me have a try with the latest qemu. I think I hold the update for some issue with qemu and need to fix them all before make the update.

File: src/platform/haswell/haswell.x.in [hunk 3]
```diff
@@ -15,13 +15,13 @@ OUTPUT_ARCH(xtensa)
 MEMORY
 {
   vector_reset_text :
-        org = XCHAL_RESET_VECTOR0_PADDR,
+        org = XCHAL_RESET_VECTOR_PADDR,
```
Comment (mrajwa) [inline src/platform/haswell/haswell.x.in]: What is the reason of this change?

File: src/platform/haswell/include/platform/lib/memory.h (hunk 2 above)
Comment (mrajwa) [reply src/platform/haswell/include/platform/lib/memory.h]: But as for the alignment I believe its already done by:

```
 . = ALIGN (32);
    _virtual_thread_start = ABSOLUTE(.);
    . = . + SOF_VIRTUAL_THREAD_SIZE;
    _virtual_thread_end = ABSOLUTE(.);
```

For multi core I believe you want later one do sth like thread_ptr[core_id] right? I said "later" since as for now I don't see support for that.

So to me the size can be 0x10.

File: src/platform/haswell/haswell.x.in (hunk 3 above)
Comment (lgirdwood) [reply src/platform/haswell/haswell.x.in]: change in core-isa.h, VECTOR0 no longer exists.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@xiulipan all qemu fixes now merged (including HSW/BDW ISA updates).
